File: Scrabble-v2/web/src/app/classes/room.ts
import { Injectable } from '@angular/core';
import { Player } from './player';
import { RoomInfo } from './room-info';

@Injectable({
    providedIn: 'root',
})
export class Room {
    nbHumanPlayers: number;
    nbBots: number;
    elapsedTime: number;
    currentPlayerPseudo: string;
    players: Player[];
    roomInfo: RoomInfo;
    isBankUsable: boolean;
    isGameStarted: boolean;
    racks: string[][];

    constructor() {
        this.nbHumanPlayers = 1;
        this.nbBots = 0;
        this.roomInfo = { name: '', timerPerTurn: '', dictionary: '', gameType: '', maxPlayers: 2, nbHumans: 2, isPublic: true, pw: '' };
        this.players = [];
        this.isBankUsable = true;
        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
        for (let i = 0; i < 4; i++) {
            this.players.push(new Player());
        }
    }

    addPlayer(player: Player) {
        if (this.players.length < this.roomInfo.maxPlayers) {
            this.players.push(player);
        }
    }

    removePlayer(playerSocketId: string) {
        const playerToRemove = this.players.find((element) => element.socketId === playerSocketId);
        if (playerToRemove) {
            this.players.splice(this.players.indexOf(playerToRemove), 1);
        }
    }
}

File: Scrabble-v2/mobile/lib/models/room.dart
import 'player.dart';

class RoomInfo {
  late String name;
  late String levelGame;
  late String timerPerTurn;
  late String dictionary;
  late String gameType;
  late int maxPlayers;
  late bool isGameOver;
  late int nbHumans;
  late bool isPublic;
  late String pw;
}

class JoinInterface {
  late String roomName;
  late Player player;

  JoinInterface() {
    roomName = 'unknown';
    player = Player();
  }
}

class Room {
  late int elapsedTime;
  late String currentPlayerPseudo;
  late List<Player> players;
  late RoomInfo roomInfo;
  late bool isBankUsable;
  late bool isGameStarted;
  late int nbHumanPlayers;
  late int nbBots;

  Room() {
    roomInfo = RoomInfo();
    roomInfo.levelGame = 'easy';
    roomInfo.isPublic = true;
    players = [];
    isBankUsable = true;
    currentPlayerPseudo = "incognito";
    elapsedTime = 0;
    roomInfo.gameType = 'classic';
    isGameStarted = false;
    nbHumanPlayers = 1;
    nbBots = 0;
  }

  Map toJson() {
    Map room = {
      'elapsedTime': elapsedTime ?? 0,
      'currentPlayerPseudo': currentPlayerPseudo ?? 'incognito',
      'isBankUsable': isBankUsable ?? true,
      'isGameStarted': isGameStarted ?? false,
      'nbHumanPlayers': nbHumanPlayers ?? 1,
      'nbBots': nbBots ?? 4,
      'roomInfo': {
        'name': roomInfo.name ?? '',
        'timerPerTurn': roomInfo.timerPerTurn ?? '60',
        'dictionary': roomInfo.dictionary ?? 'dictionnaire par défaut',
        'gameType': roomInfo.gameType ?? 'classic',
        'isGameOver': roomInfo.isGameOver ?? true,
        'nbHumans': roomInfo.nbHumans ?? 2,
        'isPublic': roomInfo.isPublic ?? false,
        'pw': roomInfo.pw ?? '',
      },
      'players': []
    };

    for (var player in players) {
      room['players'].add(player.toJson());
    }

    return room;
  }
}
